src/eventListeners/token.ts

```diff
@@ -1,18 +1,41 @@
 import { ContractEventsSignatures } from '~types';
-import { addEventListener } from '~eventListeners';
-import { EventListenerType } from './types';
+import { addEventListener, getEventListeners } from '~eventListeners';
+import { EventListenerType, EventListener } from './types';
 import { utils } from 'ethers';
 
-/**
- * @NOTE: Currently, token event listeners only support filtering on the recipient address
- */
 export const addTokenEventListener = (
   eventSignature: ContractEventsSignatures,
   recipientAddress: string,
 ): void => {
-  addEventListener({
+  /*
+   * @NOTE This only applies to contract listeners
+   *
+   * Ie: any listener that has and address property.
+   * As a general rule, this will only *NOT* apply to the token Transfer event
+   * as that's the only one treated differently.
+   */
+  const listenerExists = getEventListeners().some(
+    (listener) =>
+      listener.type === EventListenerType.Token &&
+      listener.address === recipientAddress,
+  );
+
+  const tokenListener: EventListener = {
     type: EventListenerType.Token,
     eventSignature,
-    topics: [utils.id(eventSignature), null, recipientAddress],
-  });
+    address: recipientAddress,
```

Comment:
> The address property here should be set to tokenAddress, and the if below could depend on whether recipientAddress is defined, as opposed to event signature

That's not correct approach, since then you can break this by just passing whatever in in the `recipientAddress`. The "filter by recipient" patter works only in some cases, and we need to correctly control those. 

But overall I think this has become too complicated and I have half a mind to revert my changes back to my initial implementation: one handler for Transfer events, one handler for every other type of token event